pinot-broker/src/main/java/org/apache/pinot/broker/failuredetector/BaseExponentialBackoffRetryFailureDetector.java

```diff
@@ -0,0 +1,174 @@
+
+  protected BrokerMetrics _brokerMetrics;
+  protected long _retryInitialDelayNs;
+  protected double _retryDelayFactor;
+  protected int _maxRetries;
+  protected Thread _retryThread;
+
+  protected volatile boolean _running;
+
+  @Override
+  public void init(PinotConfiguration config, BrokerMetrics brokerMetrics) {
+    _brokerMetrics = brokerMetrics;
+    long retryInitialDelayMs = config.getProperty(Broker.FailureDetector.CONFIG_OF_RETRY_INITIAL_DELAY_MS,
+        Broker.FailureDetector.DEFAULT_RETRY_INITIAL_DELAY_MS);
+    _retryInitialDelayNs = TimeUnit.MILLISECONDS.toNanos(retryInitialDelayMs);
+    _retryDelayFactor = config.getProperty(Broker.FailureDetector.CONFIG_OF_RETRY_DELAY_FACTOR,
+        Broker.FailureDetector.DEFAULT_RETRY_DELAY_FACTOR);
+    _maxRetries =
+        config.getProperty(Broker.FailureDetector.CONFIG_OF_MAX_RETRIES, Broker.FailureDetector.DEFAULT_MAX_RETIRES);
+    _logger.info("Initialized {} with retry initial delay: {}ms, exponential backoff factor: {}, max " + "retries: {}",
+        _name, retryInitialDelayMs, _retryDelayFactor, _maxRetries);
+  }
+
+  @Override
+  public void register(Listener listener) {
+    _listeners.add(listener);
+  }
+
+  @Override
+  public void start() {
+    _logger.info("Starting {}", _name);
+    _running = true;
+
+    _retryThread = new Thread(() -> {
+      while (_running) {
+        try {
+          long earliestRetryTimeNs = System.nanoTime() + _retryInitialDelayNs;
+          for (Map.Entry<String, RetryInfo> entry : _unhealthyServerRetryInfoMap.entrySet()) {
+            RetryInfo retryInfo = entry.getValue();
+            if (System.nanoTime() > retryInfo._retryTimeNs) {
+              String instanceId = entry.getKey();
+              if (retryInfo._numRetires == _maxRetries) {
+                _logger.warn(
+                    "Unhealthy server: {} already reaches the max retries: {}, do not retry again and treat it as "
+                        + "healthy so that the listeners do not lose track of the server", instanceId, _maxRetries);
+                markServerHealthy(instanceId);
+                continue;
+              }
+              _logger.info("Retry unhealthy server: {}", instanceId);
+              for (Listener listener : _listeners) {
+                listener.retryUnhealthyServer(instanceId, this);
+              }
+              // Update the retry info
+              retryInfo._retryDelayNs = (long) (retryInfo._retryDelayNs * _retryDelayFactor);
+              retryInfo._retryTimeNs = System.nanoTime() + retryInfo._retryDelayNs;
+              retryInfo._numRetires++;
+            } else {
+              earliestRetryTimeNs = Math.min(earliestRetryTimeNs, retryInfo._retryTimeNs);
+            }
+          }
+          //noinspection BusyWait
+          Thread.sleep(TimeUnit.NANOSECONDS.toMillis(earliestRetryTimeNs - System.nanoTime()));
```

Comment:
This would better implemented by putting a task onto a delay queue and waiting on that queue

Reply:
Good point, this can also avoid scanning the map each time it wakes up. Applied the changes